core/modules/replica/StatusRequest.cc

```diff
@@ -0,0 +1,206 @@
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "awaken");
+
+    if (isAborted(ec)) { return; }
+
+    // Also ignore this event if the request expired
+    if (_state== State::FINISHED) { return; }
+
+    LOCK_GUARD;
+
+    // Serialize the Status message header and the request itself into
+    // the network buffer.
+
+    _bufferPtr->resize();
+
+    proto::ReplicationRequestHeader hdr;
+    hdr.set_id(id());
+    hdr.set_type(proto::ReplicationRequestHeader::REQUEST);
+    hdr.set_management_type(proto::ReplicationManagementRequestType::REQUEST_STATUS);
+
+    _bufferPtr->serialize(hdr);
+
+    proto::ReplicationRequestStatus message;
+    message.set_id(_targetRequestId);
+    message.set_type(_requestType);
+
+    _bufferPtr->serialize(message);
+
+    send();
+}
+
+void StatusRequestBaseM::analyze(bool success,
+                                 proto::ReplicationStatus status) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "analyze  success=" << (success ? "true" : "false"));
+    {
+        // This guard is made on behalf of an asynchronious callback fired
+        // upon a completion of the request within method send() - the only
+        // client of analyze()
+        LOCK_GUARD;
+
+        if (success) {
+
+            switch (status) {
+
+                case proto::ReplicationStatus::SUCCESS:
+                    finish(SUCCESS);
+                    break;
+
+                case proto::ReplicationStatus::QUEUED:
+                    if (_keepTracking) { wait(); }
+                    else               { finish(SERVER_QUEUED); }
+                    break;
+
+                case proto::ReplicationStatus::IN_PROGRESS:
+                    if (_keepTracking) { wait(); }
+                    else               { finish(SERVER_IN_PROGRESS); }
+                    break;
+
+                case proto::ReplicationStatus::IS_CANCELLING:
+                    if (_keepTracking) { wait(); }
+                    else               { finish(SERVER_IS_CANCELLING); }
+                    break;
+
+                case proto::ReplicationStatus::BAD:
+                    finish(SERVER_BAD);
+                    break;
+
+                case proto::ReplicationStatus::FAILED:
+                    finish(SERVER_ERROR);
+                    break;
+
+                case proto::ReplicationStatus::CANCELLED:
+                    finish(SERVER_CANCELLED);
+                    break;
+
+                default:
+                    throw std::logic_error(
+                            "StatusRequestBaseM::analyze() unknown status '"
+                            + proto::ReplicationStatus_Name(status) +
+                            "' received from server");
+            }
+
+        } else {
+            finish(CLIENT_ERROR);
+        }
+    }
+
+    // The client callback is made in the lock-free zone to avoid possible
+    // deadlocks
+    if (_state == State::FINISHED) { notify(); }
```

Comment:
would it be safer to decide if it should notify while holding the lock, release the lock, and then notify `if (doNotify == true)`? IE is it safe to be using the state for logic without the lock, or can the state not be set by another thread?

Reply:
This problem is now gone because a call to notify() has been moved into finish().